presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/KDBTree.java

```diff
@@ -0,0 +1,344 @@
+        checkArgument(maxLevels > 0, "maxLevels must be > 0");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+        return new KDBTree(buildKDBTreeNode(maxItemsPerNode, maxLevels, 0, extent, items, new LeafIdAllocator()));
+    }
+
+    private static Node buildKDBTreeNode(int maxItemsPerNode, int maxLevels, int level, Rectangle extent, List<Rectangle> items, LeafIdAllocator leafIdAllocator)
+    {
+        checkArgument(maxItemsPerNode > 0, "maxItemsPerNode must be > 0");
+        checkArgument(maxLevels > 0, "maxLevels must be > 0");
+        checkArgument(level >= 0, "level must be >= 0");
+        requireNonNull(extent, "extent is null");
+        requireNonNull(items, "items is null");
+
+        if (items.size() <= maxItemsPerNode || level >= maxLevels) {
+            return new Node(extent, leafIdAllocator.next());
+        }
+
+        // Split over longer side
+        boolean splitVertically = extent.getWidth() >= extent.getHeight();
+        Optional<SplitResult<Node>> splitResult = trySplit(splitVertically, maxItemsPerNode, maxLevels, level, extent, items, leafIdAllocator);
+        if (!splitResult.isPresent()) {
+            // Try spitting by the other side
+            splitResult = trySplit(!splitVertically, maxItemsPerNode, maxLevels, level, extent, items, leafIdAllocator);
+        }
+
+        if (!splitResult.isPresent()) {
+            return new Node(extent, leafIdAllocator.next());
+        }
+
+        return new Node(extent, splitResult.get().getLeft(), splitResult.get().getRight());
+    }
+
+    private static final class SplitResult<T>
+    {
+        private final T left;
+        private final T right;
+
+        private SplitResult(T left, T right)
+        {
+            this.left = requireNonNull(left, "left is null");
+            this.right = requireNonNull(right, "right is null");
+        }
+
+        public T getLeft()
+        {
+            return left;
+        }
+
+        public T getRight()
+        {
+            return right;
+        }
+    }
+
+    private static Optional<SplitResult<Node>> trySplit(boolean splitVertically, int maxItemsPerNode, int maxLevels, int level, Rectangle extent, List<Rectangle> items, LeafIdAllocator leafIdAllocator)
+    {
+        if (items.size() <= 1) {
+            // Not enough envelopes. Can't split.
+            return Optional.empty();
+        }
+
+        // Sort envelopes by xMin or yMin
+        List<Rectangle> sortedItems = ImmutableList.sortedCopyOf(splitVertically ? BY_X : BY_Y, items);
+
+        // Find a mid-point
+        int middleIndex = sortedItems.size() / 2 - 1;
+        Rectangle middleEnvelope = sortedItems.get(middleIndex);
+        double splitValue = splitVertically ? middleEnvelope.getXMin() : middleEnvelope.getYMin();
+        int splitIndex = middleIndex;
+        for (int i = middleIndex + 1; i < sortedItems.size(); i++) {
```

Comment:
I'm not sure I understand the purpose of this loop. Can you explain the intent?

Reply:
@martint In case when there are a bunch of duplicate elements in the middle of the list, this loop advances `middleIndex` past the last duplicate element. As a result, [0, middleIndex] elements fall into the left partition and [middleIndex, size) fall into the right partition.

Reply:
In that case, couldn't that be simplified to:

```java
// skip over duplicate values
while (splitIndex < sortedItems.size() && splitDomain.getValue(sortedItems.get(splitIndex)) == splitValue) {
    splitIndex++;
}

// all values in partition are the same, so can't split
if (splitIndex == sortedItems.size()) {
    return Optional.empty();
}

splitValue = (splitValue + splitDomain.getValue(sortedItems.get(splitIndex))) / 2;
```

I don't understand the last computation, though (`splitValue = (splitValue + value) / 2`). Don't you want the splitValue to match the chosen splitIndex? (vs the average between that point an the initial middle point?)

Reply:
@martint I think splitting in between values works better. E.g. splitting `(0, 1, 2, 3)` at `1.5` visually makes more sense, then splitting at `1`. Also, since we are splitting based on a sample of data, it is reasonable to assume that there are values in between and not attributed all of them to one partition.

Reply:
@martint I like the rewrite your proposed. I noticed that you used `splitDomain`. Should I assume that you think it works better than existing code and merge `SplitDomain` commit with `KdbTree` commit?

Reply:
>  Should I assume that you think it works better than existing code and merge SplitDomain

Yes, I think the code is clearer that way (fewer extraneous ifs and conditionals). Not sure about the name, though. Maybe SplitDimension? SplitAxis?

Reply:
> I think splitting in between values works better. E.g. splitting (0, 1, 2, 3) at 1.5 visually makes more sense, then splitting at 1. Also, since we are splitting based on a sample of data, it is reasonable to assume that there are values in between and not attributed all of them to one partition.

In that case, would it make sense to take into account the min *and* max values when splitting? (i.e., both edges, not just the left-most or bottom-most coordinate for each rectangle)

Reply:
@martint Martin, using a center point of a rectangle for splitting would be fine too. I don't have enough data to see if that would make any practical difference, though.